File: Remove Duplicates from Sorted List II/Remove_Duplicates_from_Sorted_List_II.py
# 从测试用例来看，貌似升序和降序都是存在的
# 60ms 49.02%
class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        pre = dummy = ListNode(0)
        while head:
            if head.next and head.val == head.next.val:
                while head and head.next and head.val == head.next.val:
                    head = head.next
                head = head.next
                if not head:
                    pre.next = head
            else:
                pre.next = head
                pre = head
                head = head.next
        return dummy.next

# 下面是同样的思路，但是更加高效的实现
# 52ms 80.79%
class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = pre = ListNode(0)
        dummy.next = head
        while head and head.next:
            if head.val == head.next.val:
                while head and head.next and head.val == head.next.val:
                    head = head.next
                head = head.next
                pre.next = head
            else:
                pre = pre.next
                head = head.next
        return dummy.next

File: Remove Duplicates from Sorted List II/Remove_Duplicates_from_Sorted_List_II.cpp
// Runtime: 12 ms, faster than 56.22% of C++ online submissions for Remove Duplicates from Sorted List II.
// Memory Usage: 9.4 MB, less than 9.71% of C++ online submissions for Remove Duplicates from Sorted List II.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode* deleteDuplicates(ListNode* head) 
    {
        ListNode dummyHead(0);
        dummyHead->next = head;

        ListNode* curNode = head;
        ListNode* preNode = &dummyHead;

        while(curNode)
        {
            if (curNode->next && curNode->val != curNode->next->val)    
            {
                preNode = curNode;
                curNode = curNode->next;
            }
            else
            {
                if (curNode->next == nullptr)
                    break;
                else
                {
                    int delValue = curNode->val;
                    while (curNode && curNode->val == delValue)
                    {
                        preNode->next = curNode->next;
                        delete curNode;
                        curNode = preNode->next;
                    }
                }
            }
        }
        return dummyHead.next;
    }
};